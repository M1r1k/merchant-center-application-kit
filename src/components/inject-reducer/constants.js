// eslint-disable-next-line import/prefer-default-export
export const ACTIVATE_PLUGIN = 'ACTIVATE_PLUGIN';
