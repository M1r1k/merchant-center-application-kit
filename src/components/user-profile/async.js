import Loadable from 'react-loadable';
import ApplicationLoader from '../application-loader';

export default Loadable({
  loader: () => import('./user-profile'),
  loading: ApplicationLoader,
});
