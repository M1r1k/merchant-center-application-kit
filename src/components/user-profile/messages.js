import { defineMessages } from 'react-intl';

export default defineMessages({
  subtitle: {
    id: 'UserProfile.subtitle',
    description:
      'Text to describe how many organizations and projects the user have',
    defaultMessage:
      'You are in {organizationsCount, plural, one {{organizationsCount} organization} other {{organizationsCount} organizations}} and working on {projectsCount, plural, one {{projectsCount} project} other {{projectsCount} projects}}',
  },
  userUpdated: {
    id: 'UserProfile.userUpdated',
    description: 'Notification message for successful update',
    defaultMessage: 'Your profile has been successfully updated',
  },
});
