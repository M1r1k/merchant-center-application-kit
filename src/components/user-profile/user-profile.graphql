mutation UpdateUserProfile($version: Int!, $user: UserProfileInput!) {
  updateUserProfile(version: $version, user: $user) {
    id
    version
    email
    firstName
    lastName
    language
    timeZone
  }
}
