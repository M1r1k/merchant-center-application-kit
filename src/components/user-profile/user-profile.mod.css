@import '@commercetools-local/ui-kit/materials/colors.mod.css';

.header {
  background-color: var(--color-whiteSnow);
  border-bottom: 1px var(--color-greyMischka) solid;
  word-break: break-all;
}
