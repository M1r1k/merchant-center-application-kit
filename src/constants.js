// eslint-disable-next-line import/prefer-default-export
export const INTERCOM_TRACKING_STATUS = {
  active: 'ACTIVE',
  inactive: 'INACTIVE',
  pending: 'PENDING',
};
