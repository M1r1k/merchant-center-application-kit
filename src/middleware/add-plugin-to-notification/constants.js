export const __LOCAL = 'LOCAL';
export const __GLOBAL = 'GLOBAL';
