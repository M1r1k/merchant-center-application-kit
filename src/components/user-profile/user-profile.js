import React from 'react';
import PropTypes from 'prop-types';
import { compose, withProps } from 'recompose';
import { FormattedMessage, injectIntl, intlShape } from 'react-intl';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import { SubmissionError } from 'redux-form';
import Spacings from '@commercetools-local/ui-kit/materials/spacings';
import Text from '@commercetools-local/ui-kit/typography/text';
import { GRAPHQL_TARGETS, DOMAINS } from '@commercetools-local/constants';
import logger from '@commercetools-local/utils/logger';
import Avatar from '@commercetools-local/core/components/avatar';
import injectConfiguration from '@commercetools-local/core/components/configuration/inject-configuration';
import UserProfileForm from '../user-profile-form';
import styles from './user-profile.mod.css';
import messages from './messages';

// TODO: move it to uikit?
const Header = props => <div className={styles.header}>{props.children}</div>;
Header.displayName = 'Header';
Header.propTypes = { children: PropTypes.node.isRequired };

const UserProfileMutation = gql`
  mutation($version: Int!, $user: UserProfileInput!) {
    updateUserProfile(version: $version, user: $user) {
      id
      version
      email
      firstName
      lastName
      language
      timeZone
    }
  }
`;

export class UserProfile extends React.Component {
  static displayName = 'UserProfile';

  static propTypes = {
    route: PropTypes.object.isRequired,
    user: PropTypes.shape({
      firstName: PropTypes.string.isRequired,
      lastName: PropTypes.string.isRequired,
      email: PropTypes.string.isRequired,
    }),
    projectsCount: PropTypes.number.isRequired,
    organizationsCount: PropTypes.number.isRequired,

    updateUserProfile: PropTypes.func.isRequired,
    showNotification: PropTypes.func.isRequired,
    showUnexpectedErrorNotification: PropTypes.func.isRequired,

    env: PropTypes.string.isRequired,
    intl: intlShape.isRequired,
  };

  static defaultProps = {
    projectsCount: 0,
    organizationsCount: 0,
  };

  handleSave = formData =>
    this.props
      .updateUserProfile({
        variables: {
          target: GRAPHQL_TARGETS.MerchantCenterBackend,
          version: formData.version,
          user: {
            firstName: formData.firstName,
            lastName: formData.lastName,
            language: formData.language,
            timeZone: formData.timeZone,
          },
        },
      })
      .then(() => {
        // Dispatch a success notification
        this.props.showNotification({
          kind: 'success',
          domain: DOMAINS.SIDE,
          text: this.props.intl.formatMessage(messages.userUpdated),
        });
      })
      .catch(error => {
        // On production we send the errors to Sentry.
        if (this.props.env !== 'production') logger.error(error, error.stack);
        // Show an error message
        this.props.showUnexpectedErrorNotification({ error });
        // Notify redux-form that there was an error.
        throw new SubmissionError();
      });

  render() {
    if (!this.props.user) return null;
    return (
      <div
        /* TODO: build this into UIKit (e.g. <Page>) */
        style={{ height: '100%', display: 'flex', flexDirection: 'column' }}
      >
        <Header>
          <Spacings.Inset>
            <Spacings.Inline scale="m" alignItems="center">
              <Avatar
                email={this.props.user.email}
                firstName={this.props.user.firstName}
                lastName={this.props.user.lastName}
                size="big"
              />
              <div>
                <Text.Headline elementType="h1">
                  {`${this.props.user.firstName} ${this.props.user.lastName}`}
                </Text.Headline>
                <Text.Body>
                  <FormattedMessage
                    {...messages.subtitle}
                    values={{
                      organizationsCount: this.props.organizationsCount,
                      projectsCount: this.props.projectsCount,
                    }}
                  />
                </Text.Body>
              </div>
            </Spacings.Inline>
          </Spacings.Inset>
        </Header>

        <div
          /* TODO: build this into UIKit (e.g. <Page.Content scrollable={true}>) */
          style={{ overflow: 'auto', flexGrow: 1 }}
        >
          <UserProfileForm
            // This is specific to redux-form
            initialValues={this.props.user}
            onSubmit={this.handleSave}
            route={this.props.route}
          />
        </div>
      </div>
    );
  }
}

export default compose(
  injectIntl,
  withProps(props => ({
    projectsCount: props.user ? props.user.availableProjects.length : 0,
    organizationsCount: props.user
      ? props.user.availableOrganizations.length
      : 0,
  })),
  graphql(UserProfileMutation, { name: 'updateUserProfile' }),
  injectConfiguration('env', 'env')
)(UserProfile);
