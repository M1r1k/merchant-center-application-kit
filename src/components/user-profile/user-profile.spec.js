import React from 'react';
import { shallow } from 'enzyme';
import { GRAPHQL_TARGETS, DOMAINS } from '@commercetools-local/constants';
import { intlMock } from '@commercetools-local/test-utils';
import UserProfileForm from '../user-profile-form';
import { UserProfile } from './user-profile';

const createTestProps = props => ({
  route: {},
  user: {
    firstName: 'foo',
    lastName: 'bar',
    email: 'foo@bar.com',
  },
  projectsCount: 2,
  organizationsCount: 1,
  updateUserProfile: jest.fn(),
  showNotification: jest.fn(),
  showApiErrorNotification: jest.fn(),
  intl: intlMock,
  ...props,
});

describe('rendering', () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<UserProfile {...props} />);
  });
  it('should ensure layout structure', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should render <Avatar>', () => {
    expect(wrapper).toRender('Avatar');
  });
  it('should render <TextHeadline>', () => {
    expect(wrapper).toRender('TextHeadline');
  });
  it('should render <TextBody>', () => {
    expect(wrapper).toRender('TextBody');
  });
  it('should render <TextBody> message', () => {
    expect(wrapper.find('TextBody')).toRender({ id: 'UserProfile.subtitle' });
  });
  it('should pass values to <TextBody> message', () => {
    expect(wrapper.find('TextBody')).toRender({
      values: { organizationsCount: 1, projectsCount: 2 },
    });
  });
  describe('<UserProfileForm>', () => {
    it('should pass initialValues to <UserProfileForm>', () => {
      expect(wrapper.find(UserProfileForm)).toHaveProp(
        'initialValues',
        props.user
      );
    });
  });
});

describe('callbacks', () => {
  describe('handleSubmit', () => {
    let props;
    let wrapper;
    let formikBag;
    describe('when request is successfull', () => {
      beforeEach(() => {
        formikBag = {
          setSubmitting: jest.fn(),
          resetForm: jest.fn(),
        };
        props = createTestProps({
          updateUserProfile: jest.fn(() =>
            Promise.resolve({
              data: {
                updateUserProfile: {
                  id: '1',
                  version: 2,
                  email: 'john@snow.got',
                  firstName: 'John',
                  lastName: 'Snow',
                  language: 'en',
                  timeZone: 'Europe/Berlin',
                },
              },
            })
          ),
        });
        wrapper = shallow(<UserProfile {...props} />);
        return wrapper.instance().handleSubmit(
          {
            version: 2,
            firstName: 'John',
            lastName: 'Snow',
            language: 'en',
            timeZone: 'Europe/Berlin',
          },
          formikBag
        );
      });
      it('should trigger mutation', () => {
        expect(props.updateUserProfile).toHaveBeenCalledTimes(1);
        expect(props.updateUserProfile).toHaveBeenCalledWith({
          variables: {
            target: GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND,
            version: 2,
            user: {
              firstName: 'John',
              lastName: 'Snow',
              language: 'en',
              timeZone: 'Europe/Berlin',
            },
          },
        });
      });
      it('should dispatch success notification', () => {
        expect(props.showNotification).toHaveBeenCalledTimes(1);
        expect(props.showNotification).toHaveBeenCalledWith({
          kind: 'success',
          domain: DOMAINS.SIDE,
          text: 'UserProfile.userUpdated',
        });
      });
      it('should reset form values', () => {
        expect(formikBag.resetForm).toHaveBeenCalled();
      });
      it('should mark form as not submitting', () => {
        expect(formikBag.setSubmitting).toHaveBeenCalledWith(false);
      });
    });
    describe('when request fails', () => {
      beforeEach(() => {
        formikBag = {
          setSubmitting: jest.fn(),
          setErrors: jest.fn(),
        };
        props = createTestProps({
          updateUserProfile: jest.fn(() => Promise.reject(new Error('Oops'))),
        });
        wrapper = shallow(<UserProfile {...props} />);
        return wrapper.instance().handleSubmit(
          {
            version: 2,
            firstName: 'John',
            lastName: 'Snow',
            language: 'en',
            timeZone: 'Europe/Berlin',
          },
          formikBag
        );
      });
      it('should trigger mutation', () => {
        expect(props.updateUserProfile).toHaveBeenCalledTimes(1);
        expect(props.updateUserProfile).toHaveBeenCalledWith({
          variables: {
            target: GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND,
            version: 2,
            user: {
              firstName: 'John',
              lastName: 'Snow',
              language: 'en',
              timeZone: 'Europe/Berlin',
            },
          },
        });
      });
      it('should not dispatch success notification', () => {
        expect(props.showNotification).toHaveBeenCalledTimes(0);
      });
      it('should mark form as not submitting', () => {
        expect(formikBag.setSubmitting).toHaveBeenCalledWith(false);
      });
    });
  });
});
