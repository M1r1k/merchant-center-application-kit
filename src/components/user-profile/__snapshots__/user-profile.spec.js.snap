// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should ensure layout structure 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
    }
  }
>
  <Header>
    <Inset
      scale="m"
    >
      <Inline
        alignItems="center"
        scale="m"
      >
        <Avatar
          email="foo@bar.com"
          firstName="foo"
          lastName="bar"
          size="big"
        />
        <div>
          <TextHeadline
            elementType="h1"
          >
            foo bar
          </TextHeadline>
          <TextBody>
            <FormattedMessage
              defaultMessage="You are in {organizationsCount, plural, one {{organizationsCount} organization} other {{organizationsCount} organizations}} and working on {projectsCount, plural, one {{projectsCount} project} other {{projectsCount} projects}}"
              description="Text to describe how many organizations and projects the user have"
              id="UserProfile.subtitle"
              values={
                Object {
                  "organizationsCount": 1,
                  "projectsCount": 2,
                }
              }
            />
          </TextBody>
        </div>
      </Inline>
    </Inset>
  </Header>
  <div
    style={
      Object {
        "flexGrow": 1,
        "overflow": "auto",
      }
    }
  >
    <InjectIntl(UserProfileForm)
      initialValues={
        Object {
          "email": "foo@bar.com",
          "firstName": "foo",
          "lastName": "bar",
        }
      }
      onSubmit={[Function]}
      route={Object {}}
    />
  </div>
</div>
`;
